File: June/11-06-2024/Maximum Tip Calculator.py

from typing import List


class Solution:
    def maxTip(self, n : int, x : int, y : int, arr : List[int], brr : List[int]) -> int:
  #Your Code Start
         # Calculate the differences and store them with their indices
        diff_list = []
        for i in range(n):
            diff_list.append((abs(arr[i] - brr[i]), i))
        
        # Sort the differences in descending order
        diff_list.sort(reverse=True, key=lambda x: x[0])
        
        total_tip = 0
        countA = 0
        countB = 0
        
        for diff, idx in diff_list:
            if (arr[idx] > brr[idx] and countA < x) or countB >= y:
                total_tip += arr[idx]
                countA += 1
            else:
                total_tip += brr[idx]
                countB += 1
        
        return total_tip
#your Code End



#{ 
 # Driver Code Starts
class IntArray:

    def __init__(self) -> None:
        pass

    def Input(self, n):
        arr = [int(i) for i in input().strip().split()]  #array input
        return arr

    def Print(self, arr):
        for i in arr:
            print(i, end=" ")
        print()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):

        n = int(input())

        x = int(input())

        y = int(input())

        arr = IntArray().Input(n)

        brr = IntArray().Input(n)

        obj = Solution()
        res = obj.maxTip(n, x, y, arr, brr)

        print(res)

# } Driver Code Ends

File: June/11-06-2024/Maximum Tip Calculator.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

class Array {
  public:
    template <class T>
    static void input(vector<T> &a, int n) {
        for (int i = 0; i < n; i++) {
            scanf("%d ", &a[i]);
        }
    }

    template <class T>
    static void print(vector<T> &a) {
        for (int i = 0; i < a.size(); i++) {
            cout << a[i] << " ";
        }
        cout << endl;
    }
};


// } Driver Code Ends

class Solution {
  public:
    long long maxTip(int n, int x, int y, vector<int> &arr, vector<int> &brr) {
        // Your Code Start
         vector<pair<int, int>> diff;
        for (int i = 0; i < n; ++i) {
            diff.push_back({abs(arr[i] - brr[i]), i});
        }
        sort(diff.rbegin(), diff.rend());
        
        long long totalTip = 0;
        int countA = 0, countB = 0;
        
        for (int i = 0; i < n; ++i) {
            int idx = diff[i].second;
            if ((arr[idx] > brr[idx] && countA < x) || countB >= y) {
                totalTip += arr[idx];
                ++countA;
            } else {
                totalTip += brr[idx];
                ++countB;
            }
        }
        
        return totalTip;
      //Your Code End
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    scanf("%d ", &t);
    while (t--) {

        int n;
        scanf("%d", &n);

        int x;
        scanf("%d", &x);

        int y;
        scanf("%d", &y);

        vector<int> arr(n);
        Array::input(arr, n);

        vector<int> brr(n);
        Array::input(brr, n);

        Solution obj;
        long long res = obj.maxTip(n, x, y, arr, brr);

        cout << res << endl;
    }
}

// } Driver Code Ends
